### PR TITLE
Group 5 - Extensible Calc

### PR DESCRIPTION
**Pull request by prestonr83**
- Sharma Nadakuditi
- Preston Rodriguez

### CODE REVIEW

File: calculator/exceptions.py [hunk 1]
```diff
@@ -1,7 +1,14 @@
...
+            Exception.__init__(self, 'Given operation is invalid.')
+        else:
+            Exception.__init__(self, '"{}" operation not supported.'.
+                           format(self.operation))
 
 
 class InvalidParams(Exception):
-    pass
+    def __init__(self):
+        Exception.__init__(self, 'Given params are invalid.')
```
Comment (santiagobasulto) [inline calculator/exceptions.py]: Guys, check the `super()` function.

File: calculator/main.py [hunk 2]
```diff
@@ -3,20 +3,22 @@
 from calculator.operations import *
 from calculator.exceptions import *
 
+# xyz = create_new
 
-def create_new_calculator(operations=None):
+def create_new_calculator(operations={}):
```
Comment (santiagobasulto) [inline calculator/main.py]: This is a bad idea. Subtle (and advanced) bug, but here is for you guys to read about it (Mistake 1 in this list): https://www.toptal.com/python/top-10-mistakes-that-python-programmers-make

Please remind me on Saturday to talk about this.

File: calculator/main.py [hunk 3]
```diff
@@ -26,7 +28,23 @@ def perform_operation(calc, operation, params):
     :param params: Tuple containing the list of nums to operate with.
                    ie: (1, 2, 3, 4.5, -2)
     """
-    pass
+    if operation not in get_operations(calc):
```
Comment (santiagobasulto) [inline calculator/main.py:33]: I like this use of the abstraction. Can you also remember to talk about this on Saturday? I'd like to share it with the other groups.

File: calculator/main.py [hunk 4]
```diff
@@ -26,7 +28,23 @@ def perform_operation(calc, operation, params):
...
+        raise InvalidOperation(operation)
+    
+    check_params = all(isinstance(x,(int,float)) for x in params)
```
Comment (santiagobasulto) [inline calculator/main.py]: 🙌 Pythonic! Nice!

Please use spaces to separate arguments: https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

File: calculator/main.py [hunk 5]
```diff
@@ -26,7 +28,23 @@ def perform_operation(calc, operation, params):
...
+    if not check_params and operation != 'plot':
+        raise InvalidParams()
+    
+    if not params or (len(params) == 0):
```
Comment (santiagobasulto) [inline calculator/main.py]: `not params` also include the case of an empty list/tuple.

File: calculator/main.py [hunk 6]
```diff
@@ -26,7 +28,23 @@ def perform_operation(calc, operation, params):
...
+        raise InvalidParams()
+    
+    
+    return_val = calc['operations'][operation](*params)
+    if return_val:
```
Comment (santiagobasulto) [inline calculator/main.py]: Why would `return_val` be `None`? What happens if it's `0`? (`subtract(2, 2)`)

File: calculator/main.py [hunk 7]
```diff
@@ -58,18 +80,25 @@ def get_history(calc):
...
 def repeat_last_operation(calc):
     """
     Returns the result of the last operation executed in the history.
     """
-    pass
+    if len(calc['history']) == 0:
+        return None
+    
+    _,_,_,return_val = calc['history'][-1]
+    return return_val
```
Comment (santiagobasulto) [inline calculator/main.py:104]: Can't this just be `return calc['history'][-1]`?

File: calculator/main.py [hunk 8]
```diff
@@ -58,18 +80,25 @@ def get_history(calc):
...
+    calc['history'] = []
+
 
 
 def repeat_last_operation(calc):
     """
     Returns the result of the last operation executed in the history.
     """
-    pass
+    if len(calc['history']) == 0:
```
Comment (santiagobasulto) [inline calculator/main.py]: Usually preferred to do:

``` python
if not calc['history']:
    # blah blah
```

File: calculator/operations.py [hunk 9]
```diff
@@ -1,23 +1,30 @@
...
-    # your implementation here
-    pass
+    return reduce(lambda x, y: x*y, args)
 
 def divide(*args):
-    # your implementation here
-    pass
+    return reduce(lambda x, y: float(x)/float(y), args)
 
 def plot(*args):
```
Comment (santiagobasulto) [inline calculator/operations.py:20]: 😱😱😱😱😱

Does this work? Can you add a screenshot?

File: calculator/exceptions.py (hunk 1 above)
Comment (prestonr83) [reply calculator/exceptions.py]: So from what I gather, super() allows you to the call base class with having to explicitly define it. IE in this case

`super(InvalidParams,self).__init__('Given params are invalid'`

Is that best practice? Is this benefit just that if the Base class changed we wouldn't have to refactor the explicit calls to it within the class? Is it wrong to do it the way we did in such a simple class or should we just always prefer to use super()?

File: calculator/main.py (hunk 7 above)
Comment (prestonr83) [reply calculator/main.py:104]: @santiagobasulto 
That would return the whole tuple not just the total of the last operation. We could have just done 

`return_val = calc['history'][-1]`
`return return_val[3]`
or just
`return calc['history'][-1][3]`

File: calculator/operations.py (hunk 9 above)
Comment (prestonr83) [reply calculator/operations.py:20]: @santiagobasulto 
Yes it does
[Plot screenshot](https://github.com/prestonr83/pyp-w1-gw-extensible-calculator/blob/master/plot_sample.png)
[Plot sample output](https://github.com/prestonr83/pyp-w1-gw-extensible-calculator/blob/master/Plot_Output_Sample.txt)

File: calculator/main.py (hunk 6 above)
Comment (prestonr83) [reply calculator/main.py]: @santiagobasulto Doh!

**Comment by charliegphan** [comment]
Clever use of reduce, good work guys! Also the main file is very clean, I like it.

**Pull request closed without merging**